### PR TITLE
Added Windows DeltaCount counter

### PR DESCRIPTION
**Pull request by monaf**
We needed only Delta type counter type as Windows Performance Counter. Graphite is not necessary for us.

### CODE REVIEW

File: src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCountCounter.cs]: This comment is incorrect. It should be rewritten in English. Something like this: "A difference counter that shows the change in the measured value between the two most recent sample intervals (for CompositeCounters)"

File: src/Qoollo.PerformanceCounters/Generic/Categories/InstanceInMultiInstanceCategory.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Categories/InstanceInMultiInstanceCategory.cs]: The name of the method should match the name of the counter (by convention). Please, change it to ```GetDeltaCountCounter```

File: src/Qoollo.PerformanceCounters/Generic/Categories/InstanceInMultiInstanceCategory.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Categories/InstanceInMultiInstanceCategory.cs]: NumberOfItemsCounter -> DeltaCountCounter

File: src/Qoollo.PerformanceCounters/Generic/Categories/MultiInstanceCategory.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Categories/MultiInstanceCategory.cs]: NumberOfItemsCounter -> DeltaCountCounter

File: src/Qoollo.PerformanceCounters/Generic/Categories/SingleInstanceCategory.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Categories/SingleInstanceCategory.cs]: NumberOfItemsCounter -> DeltaCountCounter

File: src/Qoollo.PerformanceCounters/Generic/Categories/SingleInstanceCategory.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Categories/SingleInstanceCategory.cs]: NumberOfItemsCounter -> DeltaCountCounter

File: src/Qoollo.PerformanceCounters/Generic/Counters/DeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Counters/DeltaCountCounter.cs]: The comment is incorrect.
It would be great to rewrite all comments in this file in English.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCountCounter.cs]: Again: comment is incorrect. It would be better to rewrite it in English

File: src/Qoollo.PerformanceCounters/NullCounters/Counters/NullDeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/NullCounters/Counters/NullDeltaCountCounter.cs]: Again: comment is incorrect. It would be better to rewrite it in English

File: src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCountCounter.cs]: Again: comment is incorrect. It would be better to rewrite it in English

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCountCounter.cs]: This is the most problematic place in this PR:

1. Design problem: ```CurrentValue``` access should not affect the value of the counter. It should be a readonly property. To achieve this behaviour, measurements and readings must be separated. For the measurement purposes the new method ```long Measure()``` should be added to the base class and implemented here.
2. Implementation problem: Implementation is not thread safe. It should use ```Interlocked.CompareExchange``` inside loop with ```SpinWait``` to update the value atomically.

File: src/Qoollo.PerformanceCounters/Generic/Counters/DeltaCountCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/Generic/Counters/DeltaCountCounter.cs]: New method should be added here to perform measurement: ```public abstract long Measure();```
See comment in InternalDeltaCountCounter.cs

**Review by ikopylov — CHANGES_REQUESTED** [review]
There are problems that need to be resolved before merge. The main one is in the design of DeltaCounter.

DeltaCounter support should also be added to GraphiteCounters and NetCounters, but that can be done in a separate PR.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: Please, make the definition of the property old-stylish (``` { get { return ...; }}```)

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs:135]: I have tested the behaviour of the WinCounters in Performance Counters Monitor and behaviour of this implementation and found that they are differ. The problem is highlighted when ```SetValue(long value)``` used. 
Example: 
``` while (true) { deltaCounter.SetValue(100); Thread.Sleep(1000); Console.WriteLine(deltaCounter.Measure()); } ```
Performance Counters Monitor correctly displays '0' while this implementation prints '100'.

The solution for this problem would be to use the original idea to store ```_lastMeasuredValue``` in the field. During the testing I made the proper implementation (😄) and it looks like this:
```
        public override long Measure()
        {
            long currentValue = Interlocked.Read(ref _currentValue);
            long lastMeasuredValue = Interlocked.Exchange(ref _lastMeasuredValue, currentValue);
            return currentValue - lastMeasuredValue;
        }
```

Could you simply change the code to the listed above

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: _Side notes_:
The scheme you implemented is thread-safe, but in this situation it is unnecessarily complicated. I've noticed that you took the scheme inside the ```InternalAverageCountCounter``` as the basis, but there it was used for atomic reading of __two__  dependent values. When only one value has to be read/updated atomically, everything can be done using methods from ```Interlocked``` class.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: IMO, ```CurrentValue``` should work the same way as ```Measure``` except that it should not update ```_lastMeasuredValue``` field.
```
       public override long CurrentValue 
        { 
            get 
            {
                long currentValue = Interlocked.Read(ref _currentValue);
                long lastMeasuredValue = Interlocked.Read(ref _lastMeasuredValue);
                return currentValue - lastMeasuredValue;
            } 
        }
```

File: src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs]: This will return the raw counter value (not the delta from the previous measurement).
The simpliest solution would be to do the same calculations as in ```InternalDeltaCounter```: 
```
        public override long CurrentValue
        {
            get
            {
                var counterCpy = _winCounter;
                if (counterCpy == null)
                    return Counter.FailureNum;


                long currentValue = counterCpy.RawValue;
                long lastMeasuredValue = Interlocked.Read(ref _lastMeasuredValue);
                return currentValue - lastMeasuredValue;
            }
        }
```

```_lastMeasuredValue``` field should be added to the class and should be updated inside ```Measure``` method.

File: src/Qoollo.PerformanceCounters/NullCounters/Counters/NullDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/NullCounters/Counters/NullDeltaCounter.cs]: Please, add the xml-comment to the method (the same as in the base class) to prevent the compilation warning

File: src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCounter.cs]: Please, add the xml-comment to the method (the same as in the base class) to prevent the warning during compilation

**Review by ikopylov — CHANGES_REQUESTED** [review]

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs:135]: I checked my solution one more time and found that it has one small problem. 
Suppose there is only one thread that updates counter in a cycle by calling ```Increment()```. In this situation any call to ```Measure()``` should never return negative value, but in my implementation it can.
Scenario:
1. Thread 1 enters ```Measure()``` and read ```_currentValue``` == __X__
2. Updating thread calls ```Increment()```
3. Thread 2 enters ```Measure()``` and read ```_currentValue``` == __X + 1__
4. Thread 2 reads and updates atomially ```_lastMeasuredValue``` so it becomes __X + 1__ and returns the correct value
5. Thread 1 reads and updates atomially ```_lastMeasuredValue``` and observe that it is equal to __X + 1__ whereas ```_currentValue``` was read equal to __X__. As a result the thread will return __-1__

To prevent such scenario all threads that call ```Measure()``` should be synchronized with lightweight lock that has already been implemented by you.

So the correct implementation would be:
```C#
public override long Measure()
{
    long result = 0;
    try { }
    finally 
    {
        var sw = new SpinWait();
        while (!TryTakeLock())
            sw.SpinOnce();

        long currentValue = Interlocked.Read(ref _currentValue);
        long lastMeasuredValue = Interlocked.Exchange(ref _lastMeasuredValue, currentValue);
        result = currentValue - lastMeasuredValue;

        ReleaseLock();
    }
    return result;
}
```

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (ikopylov) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: In the scenario described below (in my comment to ```Measure()``` method) ```CurrentValue``` can also return negative value. But here it can be fixed simply be reordering reads of ```_currentValue``` and ```_lastMeasuredValue```:

```C#
       public override long CurrentValue 
        { 
            get 
            {
                long lastMeasuredValue = Interlocked.Read(ref _lastMeasuredValue);
                long currentValue = Interlocked.Read(ref _currentValue);
                return currentValue - lastMeasuredValue;
            } 
        }
```

File: src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs
Comment (ikopylov) [reply src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs]: The same reodering update as in ```InternalDeltaCounter``` required here:
```C#
        public override long CurrentValue
        {
            get
            {
                var counterCpy = _winCounter;
                if (counterCpy == null)
                    return Counter.FailureNum;

                long lastMeasuredValue = Interlocked.Read(ref _lastMeasuredValue);
                long currentValue = counterCpy.RawValue;
                return currentValue - lastMeasuredValue;
            }
        }
```

File: src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCounter.cs
Comment (monaf) [reply src/Qoollo.PerformanceCounters/CompositeCounters/Counters/CompositeDeltaCounter.cs]: Ok, added.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (monaf) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: Ok.

Comment (monaf) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: Sure, in the case we would have only one value, there would be no need to have the same implementation. Now it is understandable.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (monaf) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs:135]: Sure, it was my thinking also, so I just wanted to lock those statements to critical section. Using your code.

File: src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs
Comment (monaf) [reply src/Qoollo.PerformanceCounters/InternalCounters/Counters/InternalDeltaCounter.cs]: You are right, I got the feeling it is too complicated, but I thought there had to be other reason. Anyway, I believe method GetValueAtomic does not have to be there.

File: src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs
Comment (monaf) [reply src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs]: ok.

File: src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs
Comment (ikopylov) [inline src/Qoollo.PerformanceCounters/WinCounters/Counters/WinDeltaCounter.cs]: ```NextValue()``` returns difference (delta) between two measurements. To calculate ```CurrentValue``` correctly it is required to store ```RawValue``` into ```_lastMeasuredValue``` field. And it is better to do this before the call to ```counterCpy.NextValue()``` to prevent ordering problems.

**Review by ikopylov — APPROVED** [review]

**Comment by ikopylov** [comment]
@monaf Thank you for useful addition!
I will test everything one more time in the evening and will merge PR, if there is no problems.

**Comment by monaf** [comment]
Thank you very much, it was really useful for me also to learn about "heavy" concurrent library usage.

**Pull request merged**
